tools/clang/tools/dxcompiler/dxcpdbutils.cpp

```diff
@@ -390,28 +414,66 @@ struct DxcPdbUtils : public IDxcPdbUtils, public IDxcPixDxilDebugInfoFactory
         const hlsl::DxilCompilerVersion *header = (const hlsl::DxilCompilerVersion *)(part+1);
         m_VersionInfo = *header;
         m_HasVersionInfo = true;
+
+        const char *ptr = (const char *)(header+1);
+        unsigned commitShaLength = 0;
+        unsigned i = 0;
+
+        const char *commitSha = (const char *)(header+1) + i;
+        for (; i < header->VersionStringListSizeInBytes; i++) {
+          if (ptr[i] == 0) {
+            commitShaLength = i;
+            i++;
+            break;
+          }
+        }
+
+        const char *versionString = (const char *)(header+1) + i;
+        unsigned versionStringLength = 0;
+        for (; i < header->VersionStringListSizeInBytes; i++) {
+          if (ptr[i] == 0) {
+            commitShaLength = i;
+            i++;
+            break;
+          }
+        }
+
+        m_VersionCommitSha.assign(commitSha, commitShaLength);
+        m_VersionString.assign(versionString, versionStringLength);
+
       } break;
 
       case hlsl::DFCC_ShaderSourceInfo:
       {
         const hlsl::DxilSourceInfo *header = (const hlsl::DxilSourceInfo *)(part+1);
         hlsl::SourceInfoReader reader;
-        reader.Read(header);
+        if (!reader.Init(header, part->PartSize)) {
+          Reset();
+          return E_FAIL;
+        }
 
-        m_TargetProfile = ToWstring(reader.GetTargetProfile());
-        m_EntryPoint = ToWstring(reader.GetEntryPoint());
+        // Args
+        for (unsigned i = 0; i < reader.GetArgPairCount(); i++) {
+          const hlsl::SourceInfoReader::ArgPair &pair = reader.GetArgPair(i);
+          bool excludeFromFlags = false;
+          if (pair.Name == "E") {
+            m_EntryPoint = ToWstring(pair.Value);
+            excludeFromFlags = true;
+          }
+          else if (pair.Name == "T") {
+            m_TargetProfile = ToWstring(pair.Value);
+            excludeFromFlags = true;
+          }
+          else if (pair.Name == "D") {
+            m_Defines.push_back(ToWstring(pair.Value));
+            excludeFromFlags = true;
+          }
 
-        // Defines
-        {
-          StringRef defines_data = reader.GetDefines();
-          ReadNullSeparatedStringList(defines_data, &m_Defines);
-        }
+          std::wstring rendered = ToWstring(pair.Render());
+          if (!excludeFromFlags)
+            m_Flags.push_back(rendered);
 
-        // Args and Flags
-        {
-          StringRef args_data = reader.GetArgs();
-          ReadNullSeparatedStringList(args_data, &m_Args);
-          m_Flags = ComputeFlagsBasedOnArgs(m_Args);
+          m_Args.push_back(rendered);
```

Comment:
I don't think pushing both option and value in one argument is right for all cases (see comment pointing out that not all options support joined form).